File: Head_First_Design_Patterns/src/abstract_factory/python/Pizza.py
# -*- coding: utf-8 -*-
"""
Created on Fri Oct 19 21:36:47 2018

@author: 凯风
"""

from abc import ABCMeta, abstractmethod

class Pizza(metaclass = ABCMeta):
    name = ""
    dough = None
    sauce = None
    veggies = []
    cheese = None
    pepperoni = None
    clam = None
    
    @abstractmethod
    def prepare(self):
        pass
    
    def bake(self):
        print("Bake for 25 minutes at 350")
        
    def cut(self):
        print("Cutting")
    
    def box(self):
        print("boxing")
    
    def getName(self):
        return self.name
    
    def setName(self, name):
        self.name = name

    def __str__(self):
        result = "---- " + self.name + " ----\n"
        if (self.dough != None):
            result += str(self.dough) + "\n"
        if (self.sauce != None):
            result += str(self.sauce) + "\n"
        if (self.cheese != None):
            result += str(self.cheese) + "\n"
        if (self.veggies.count != 0):
            for i in range(len(self.veggies)):
                result += str(self.veggies[i])
                if (i < len(self.veggies) -1):
                    result += ", "
        if (self.clam != None):
            result += str(self.clam) + "\n"
        
        if (self.pepperoni != None):
            result += str(self.pepperoni) + "\n"
        
        return result
        
        

class CheesePizza(Pizza):
    def __init__(self, ingredientFactory):
        self.ingredientFactory = ingredientFactory
    
    def prepare(self):
        print("Prepareing " + self.name)
        self.dough = self.ingredientFactory.createDough()
        self.sauce = self.ingredientFactory.createSauce()
        self.cheese = self.ingredientFactory.createCheese()
        

class VeggiePizza(Pizza):
    def __init__(self, ingredientFactory):
        self.ingredientFactory = ingredientFactory
    
    def prepare(self):
        print("Prepareing " + self.name)
        self.dough = self.ingredientFactory.createDough()
        self.sauce = self.ingredientFactory.createSauce()
        self.cheese = self.ingredientFactory.createCheese()
        self.veggies = self.ingredientFactory.createVeggies()


class ClamPizza(Pizza):
    def __init__(self, ingredientFactory):
        self.ingredientFactory = ingredientFactory
    
    def prepare(self):
        print("Prepareing " + self.name)
        self.dough = self.ingredientFactory.createDough()
        self.sauce = self.ingredientFactory.createSauce()
        self.cheese = self.ingredientFactory.createCheese()
        self.clam = self.ingredientFactory.createClam()


class PepperoniPizza(Pizza):
    def __init__(self, ingredientFactory):
        self.ingredientFactory = ingredientFactory
    
    def prepare(self):
        print("Prepareing " + self.name)
        self.dough = self.ingredientFactory.createDough()
        self.sauce = self.ingredientFactory.createSauce()
        self.cheese = self.ingredientFactory.createCheese()
        self.veggies = self.ingredientFactory.createVeggies()
        self.pepperoni = self.ingredientFactory.createPepperoni()

File: Head_First_Design_Patterns/src/abstract_factory/Pizza.java
package abstract_factory;

public abstract class Pizza {
    String name;
    Dough dough;
    Sauce sauce;
    Veggies veggies[];
    Cheese cheese;
    Pepperoni pepperoni;
    Clams clams;

    abstract void prepare();

    void bake(){
        System.out.println("Bake for 25min at 350");
    }

    void cut(){
        System.out.println("Cutting the pizza into diagonal slices");
    }

    void box(){
        System.out.println("Place pizza in official PizzaStore box");
    }

    void setName(String name){
        this.name = name;
    }

    String getName(){
        return name;
    }

    public String toString(){
        StringBuffer result = new StringBuffer();
        result.append("---- " + name + " ----\n");
        if (dough != null)
            result.append(dough + "\n");
        if (sauce != null)
            result.append(sauce + "\n");
        if (cheese != null)
            result.append(cheese + "\n");

        if (veggies != null){
            for (int i = 0; i < veggies.length; i++){
                result.append(veggies[i]);
                if (i < veggies.length-1)
                    result.append(", ");
            }
        }

        if (clams != null)
            result.append(clams + "\n");
        if (pepperoni != null)
            result.append(pepperoni + "\n");

        return result.toString();
    }
}
